File: tennis-game/tennis-game-kotlin-arrow/src/test/src/TennisGameTest.kt
import org.junit.Assert
import org.junit.Test
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.io.PrintStream

class TennisGameTest {

    @Test
    fun aCompleteGame() {
        val inputs: String = inputs("1", "1", "1", "1")
        val result = runGameFor(inputs)

        val expected = outputs(
                "Welcome to the Tennis Game!",
                "Which player will play (1 or 2)?",
                "Player 1 Fifteen - Player 2 Love",
                "Which player will play (1 or 2)?",
                "Player 1 Thirty - Player 2 Love",
                "Which player will play (1 or 2)?",
                "Player 1 Forty - Player 2 Love",
                "Which player will play (1 or 2)?",
                "Player 1 wins"
        )

        Assert.assertEquals(result, expected)
    }

    @Test
    fun invalidPlayer() {
        val result = runGameFor(inputs("1", "1", "99", "1", "1"))

        val expected = outputs(
                "Welcome to the Tennis Game!",
                "Which player will play (1 or 2)?",
                "Player 1 Fifteen - Player 2 Love",
                "Which player will play (1 or 2)?",
                "Player 1 Thirty - Player 2 Love",
                "Which player will play (1 or 2)?",
                "Invalid player selected. Try again.",
                "Which player will play (1 or 2)?",
                "Player 1 Forty - Player 2 Love",
                "Which player will play (1 or 2)?",
                "Player 1 wins"
        )

        Assert.assertEquals(result, expected)
    }

    private fun runGameFor(inputs: String) =
        captureOutputFor(inputs) { tennisGame().unsafeRunSync() }

    private fun inputs(vararg value: String): String = value.joinToString(enter)

    private fun outputs(vararg value: String): String = value.joinToString(enter) + enter

    private fun <A> captureOutputFor(inputs: String, toExecute: () -> A): String  {
        val swapStreams = { inputStream: InputStream, printStream: PrintStream ->
            System.setIn(inputStream)
            System.setOut(printStream)
        }
        val initialOut = System.out
        val initialIn = System.`in`
        val byteArrayOutputStream = ByteArrayOutputStream()
        swapStreams(ByteArrayInputStream(inputs.toByteArray()), PrintStream(byteArrayOutputStream))
        toExecute()
        swapStreams(initialIn, initialOut)
        return byteArrayOutputStream.toString()
    }

    private val enter = System.getProperty("line.separator")

}

File: tennis-game/typescript-fpts-tennis-game/__test__/TennisGame.test.ts
import {tennisGame} from "../src/TennisGame";
jest.mock('readline');

describe('TennisGame', () => {
    it('game loop', async () => {
        console.log = jest.fn();

        await tennisGame()

        expect((console.log as jest.Mock).mock.calls[0][0]).toBe("Welcome to the Tennis Game!");
        expect((console.log as jest.Mock).mock.calls[1][0]).toBe( "Player 1 Fifteen - Player 2 Love");
        expect((console.log as jest.Mock).mock.calls[2][0]).toBe( "Player 1 Thirty - Player 2 Love");
        expect((console.log as jest.Mock).mock.calls[3][0]).toBe( "Player 1 Forty - Player 2 Love");
        expect((console.log as jest.Mock).mock.calls[4][0]).toBe( "Player 1 wins");

        await tennisGame()

        expect((console.log as jest.Mock).mock.calls[5][0]).toBe("Welcome to the Tennis Game!");
        expect((console.log as jest.Mock).mock.calls[6][0]).toBe( "Player 1 Love - Player 2 Fifteen");
        expect((console.log as jest.Mock).mock.calls[7][0]).toBe( "Player 1 Love - Player 2 Thirty");
        expect((console.log as jest.Mock).mock.calls[8][0]).toBe( "Player 1 Love - Player 2 Forty");
        expect((console.log as jest.Mock).mock.calls[9][0]).toBe( "Invalid player selected. Try again.");
        expect((console.log as jest.Mock).mock.calls[10][0]).toBe( "Player 2 wins");
    });
})
